src/openfluid/ui/waresdev/WareSrcWidgetCollection.cpp

```diff
@@ -176,42 +185,140 @@ void WareSrcWidgetCollection::onCloseWareTabRequested(int Index)
-        /* fall through */
-      default:
-        break;
-    }
+bool WareSrcWidgetCollection::requestWareTabClosing(WareSrcWidget* Ware)
+{
+  // check if a process is not running
+  if (Ware->isWareProcessRunning())
+  {
+    QMessageBox::warning(Ware,
+                         tr("Process running"),
+                         tr("Closing tab is not allowed while a configure or build process is running."),
+                         QMessageBox::Close,QMessageBox::Close);
+    return false;
+  }
+
+
+  int Choice = QMessageBox::Discard;
+
+  if (Ware->isWareModified())
+  {
+    Choice = QMessageBox::information(Ware, 
+                                      tr("Modified documents"),
+                                      tr("Documents have been modified.\n\n"
+                                          "Do you want to save changes?"),
+                                      QMessageBox::SaveAll | QMessageBox::Discard | QMessageBox::Cancel,
+                                      QMessageBox::SaveAll);
+  }
+
+  bool IsClosed = false;
+  switch (Choice)
+  {
+    case QMessageBox::SaveAll:
+      Ware->saveAllFileTabs();
+      /* fall through */
+    case QMessageBox::Discard:
+      Ware->closeAllFileTabs();
+      closeWareTab(Ware);
+      IsClosed = true;
+      break;
+    case QMessageBox::Cancel:
+      /* fall through */
+    default:
+      break;
+  }
+  return IsClosed;
+}
+
+
+// =====================================================================
+// =====================================================================
+
+
+void WareSrcWidgetCollection::onMigrationRequestedOnWare(const QString& WarePath)
+{
+  QMap<QString, WareSrcWidget*>::iterator it = m_WareSrcWidgetByPath.find(WarePath);
+  if (it == m_WareSrcWidgetByPath.end())
+  {
+    // TOIMPL add message if not found
+
+    QMessageBox::warning(nullptr,
+                         tr("Migration failure"),
+                         tr("Ware requested not found"),
+                         QMessageBox::Close,QMessageBox::Close);
+    return;
+  }
+
+  MigrationSetupDialog MigrationSetup;
+  bool Proceed = true;
+  if (!MigrationSetup.exec())
+  {
+    Proceed = false;
+  }
+
+  // try to close ware tab before migration
+  if (Proceed && !requestWareTabClosing(it.value()))
+  {
+    Proceed = false;
   }
+
+  if (!Proceed)
+  {
+
+    QMessageBox::warning(it.value(),
+                         tr("Migration failure"),
+                         tr("Migration cancelled by user."),
+                         QMessageBox::Close,QMessageBox::Close);
+    return;
+  }
+  
+  
+  openfluid::ui::waresdev::WaresSrcIOProgressDialog ProgressDialog(tr("Ware migration"), false, mp_TabWidget);
+  
+  MigrationWorker* MWorker = new MigrationWorker(WarePath, MigrationSetup.isVerbose(), MigrationSetup.isNewBranchChecked());
+
+  try
+  {
+
+    QThread* Thread = new QThread();
+
+    MWorker->moveToThread(Thread);
+
+    connect(Thread, SIGNAL(started()), MWorker, SLOT(run()));
```

Comment:
Disconnection of signal/slots needed once finished?

Reply:
Not done in any thread connect operation, I don't know the impact on performances